Add Redirect exception support

markstory opened the pull request:
An alternative to #14384. I've added a new RedirectException class that isn't coupled to Routing package and is integrated with the ErrorHandlerMiddleware which enables any middleware/controller/component to raise a RedirectException and get the intended effect.

markstory committed:
Add Redirect exception support

An alternative to #14384. I've added a new RedirectException class that
isn't coupled to Routing package and is integrated with the
ErrorHandlerMiddleware which enables any
middleware/controller/component to raise a RedirectException and get the
intended effect.

ADmad commented on src/Http/Exception/RedirectException.php:
```diff
@@ -0,0 +1,95 @@
+ */
+class RedirectException extends Exception
+{
+
```
```suggestion
```

ADmad committed:
Fix CS error

othercorey commented on src/Routing/Exception/RedirectException.php:
````diff
@@ -27,6 +27,9 @@
  * ```
+ *
+ * If you need a more general purpose redirect exception use
+ * Cake\Http\Exception\RedirectException instead of this class.
````
Isn't it confusing to have two identically named `RedirectException` classes?

dereuromark replied:
I dont think we need this 2nd exception.
The Error package already heavily relies on the Http one:
- ErrorHandler
- ErrorHandlerMiddleware
- ExceptionRenderer

ADmad replied:
How about we deprecate this exception and make RedirectRoute also throw the new Http/Exception/RedirectException?

markstory replied:
@dereuromark Using the routing class would make standalone http package usage more cumbersome. I would rather we deprecate the Routing class and have routing depend on Http as that dependency makes more sense.

dereuromark replied:
Thats fine :)

markstory replied:
I've marked the Routing\RedirectException as deprecated, but not changed any of the code as I didn't want to potentially break compatibility over a cosmetic change.

dereuromark commented on tests/TestCase/Error/Middleware/ErrorHandlerMiddlewareTest.php:
```diff
@@ -143,6 +145,55 @@ public function testHandleException()
+        $result = $middleware->process($request, $handler);
+        $this->assertInstanceOf(ResponseInterface::class, $result);
+        $this->assertEquals(302, $result->getStatusCode());
+        $this->assertEmpty('' . $result->getBody());
```
You could also (string) cast here.

dereuromark commented on tests/TestCase/Error/Middleware/ErrorHandlerMiddlewareTest.php:
```diff
@@ -143,6 +145,55 @@ public function testHandleException()
+        $request = ServerRequestFactory::fromGlobals();
+        $middleware = new ErrorHandlerMiddleware();
+        $handler = new TestRequestHandler(function () {
+            throw new RedirectException('http://example.org/login');
```
For security purposes, shouldn't we only use internal (relative) URLs?
https://github.com/cakephp/cakephp/pull/14384/files#diff-fd25dba57f71dd4309af4df2494d43e2R55
Or is that up to the user at this point?

markstory replied:
I think that is up to the user. The exceptions and redirect response don't modify the provided url at all.

markstory committed:
Use string cast.

markstory committed:
Mark routing exception as deprecated.

I don't want to change RedirectRoute or RoutingMiddleware at this time
as I don't want to break compatibility over a small value change.

ADmad commented:
`RedirectRoute` should also be updated to throw the new exception and accordingly `RoutingMiddleware::proces()` too.

Perhaps the check for redirect exception in `RoutingMiddleware` could be just done away with since `ErrorHandlerMiddleware` now checks and handles it.

dereuromark committed:
Dont use deprecation in code.

dereuromark commented:
I fixed up the remaining part.

> Perhaps the check for redirect exception in RoutingMiddleware could be just done away with since ErrorHandlerMiddleware now checks and handles it.

That part should probably be kept for the old Exception class to work.
As the new one only checks for the new `Cake\Http\Exception\RedirectException`.

So the PR as it stands now seems fine to me.
Routing handles Route one, otherwise normal ErrorHandler handles the new one.

dereuromark commented:
How do we want to adjust the tests?
$this->expectException()?
Or should we also catch the new exception in the RoutingMiddleware?

dereuromark committed:
Allow old redirect to function in RoutingMiddleware

ADmad commented on src/Routing/Middleware/RoutingMiddleware.php:
```diff
@@ -150,6 +151,11 @@ public function process(ServerRequestInterface $request, RequestHandlerInterface
                 $e->getMessage(),
```
Why do we need to catch the new exception too in routing middleware? It will be handled by errorhandler middleware isn't it?

dereuromark replied:
Then we would need to adjust the tests, as I wrote above.
We could also do that.

markstory replied:
By raising the new exception we are risking breaking any application/plugin tests that directly use the current exception class. Is such a break worth the trouble?

dereuromark replied:
This is a minor release, and affects then only tests, easy to fix with the namespace change.
So I think if it makes it cleaner and more straightforward then yes.

ADmad replied:
Another option is make `Routing\Exception\RedirectionException` an alias of `Http\Exception\RedirectException`.

dereuromark replied:
Actually, nothing breaks now the way it is now
The Routing part still catches those exceptions - and otherwise ErrorHandler jumps in.
So the current way is maybe the most pragmatic one forward.
And it can be cleaned up in 5.x.

dereuromark commented:
Are we OK with the current status, including backward compatibility for tests?
